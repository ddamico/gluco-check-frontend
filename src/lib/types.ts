import { BloodGlucoseUnits, DiabetesMetric } from "./enums";

export interface GlucoCheckUserDocument {
  defaultMetrics: DiabetesMetric[];
  glucoseUnit: BloodGlucoseUnits;
  nightscout: {
    token: string;
    url: string;
  };
}

export type SettingsFormData = {
  defaultMetrics: DiabetesMetric[];
  nightscoutUrl: string;
  nightscoutToken: string;
  glucoseUnit: BloodGlucoseUnits;
};
