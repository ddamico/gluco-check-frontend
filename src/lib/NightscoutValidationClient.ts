import { NightscoutValidationEndpointResponse } from "./types";

type NightscoutValidationClientOptions = {
  endpointUrl: string;
};

export class NightscoutValidationClient {
  private endpointUrl: string;

  constructor(options: NightscoutValidationClientOptions) {
    this.endpointUrl = options.endpointUrl;
  }

  validateResponse(
    response: any
  ): response is NightscoutValidationEndpointResponse {
    // look into a validation library, maybe joi?
    // or just do some basic checking yourself.
    return true;
  }

  async fetchValidationStatus(
    nightscoutUrl: string,
    nightscoutToken: string
  ): Promise<NightscoutValidationEndpointResponse> {
    const requestBodyRaw = JSON.stringify({
      url: nightscoutUrl,
      token: nightscoutToken,
    });
    const headers = new Headers();
    headers.append("Content-Type", "application/json; charset=utf-8");
    const response = await fetch(this.endpointUrl, {
      method: "POST",
      body: requestBodyRaw,
      redirect: "follow",
      headers,
    });
    if (this.validateResponse(response)) {
      return response;
    }
    throw new Error("Validation endpoint returned invalid response");
  }
}
