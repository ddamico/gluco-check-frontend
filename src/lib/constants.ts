export const FIRESTORE_FIELD_PATH_GLUCOSE_UNITS = "glucoseUnit";
export const FIRESTORE_FIELD_PATH_NIGHTSCOUT = "nightscout";
export const FIRESTORE_FIELD_PATH_NIGHTSCOUT_URL = `${FIRESTORE_FIELD_PATH_NIGHTSCOUT}.url`;
export const FIRESTORE_FIELD_PATH_NIGHTSCOUT_TOKEN = `${FIRESTORE_FIELD_PATH_NIGHTSCOUT}.token`;
