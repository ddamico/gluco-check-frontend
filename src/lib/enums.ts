export enum BloodGlucoseUnits {
  mgdl = "mg/dl",
  mmoll = "mmol/L",
}
