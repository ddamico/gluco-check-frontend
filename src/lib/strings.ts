export const SUBMIT_SETTINGS_FORM = "Save Settings";
