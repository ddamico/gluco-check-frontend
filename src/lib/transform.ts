import { GlucoCheckUserDocument, SettingsFormData } from "./types";

export const userSettingsFormDataToUserSettingsDocument = (
  data: SettingsFormData
): Partial<GlucoCheckUserDocument> => {
  return {
    defaultMetrics: data.defaultMetrics,
    glucoseUnit: data.glucoseUnit,
    nightscout: {
      token: data.nightscoutToken,
      url: data.nightscoutUrl,
    },
  };
};
