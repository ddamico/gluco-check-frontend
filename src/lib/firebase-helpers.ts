import { BloodGlucoseUnits } from "./enums";
import { GlucoCheckUserDocument } from "./types";

export const DEFAULT_USER_DOCUMENT: GlucoCheckUserDocument = {
  defaultPointers: ["blood sugar"],
  glucoseUnit: BloodGlucoseUnits.mgdl,
  nightscout: {
    token: '',
    url: ''
  }
};

export const getDocumentPathForUser = (user: firebase.User): string => {
  return `users/${user.email}`;
}