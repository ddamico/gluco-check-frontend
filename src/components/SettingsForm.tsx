import React, { useState } from "react";
import { useForm } from "react-hook-form";
import { BloodGlucoseUnits } from "../lib/enums";
import { SUBMIT_SETTINGS_FORM } from "../lib/strings";
import { SettingsFormData } from "../lib/types";

type SettingsFormProps = {
  nightscoutUrl: string;
  nightscoutToken: string;
  glucoseUnit: BloodGlucoseUnits;
  onSubmit: (data: SettingsFormData) => {};
};

export default function SettingsForm({
  nightscoutUrl,
  nightscoutToken,
  glucoseUnit,
  onSubmit,
}: SettingsFormProps) {
  const { register, handleSubmit, formState, watch, errors } = useForm<
    SettingsFormData
  >();
  const onFormSubmit = async (data: SettingsFormData) => {
    try {
      await onSubmit(data);
    } catch (e) {
      // TODO: probably can't just throw here, figure
      // out what hook form will do with this error
      console.log("threw on form submit");
      throw e;
    }
  };

  return (
    <>
      {formState.isSubmitting && <>Saving your settings</>}
      {!formState.isSubmitting && (
        <form onSubmit={handleSubmit(onFormSubmit)}>
          <label>
            Nightscout URL
            <input
              name="nightscoutUrl"
              defaultValue={nightscoutUrl}
              ref={register}
            />
          </label>
          <br />
          <label>
            Nightscout Token
            <input
              name="nightscoutToken"
              defaultValue={nightscoutToken}
              ref={register}
            />
          </label>
          <br />
          <label>
            Glucose Units
            <select
              name="glucoseUnit"
              defaultValue={glucoseUnit}
              ref={register({ required: true })}
            >
              {Object.entries(BloodGlucoseUnits).map(([value, key]) => (
                <option id={key} key={key}>
                  {key}
                </option>
              ))}
            </select>
          </label>
          <br />
          <button type="submit" disabled={!formState.isDirty}>
            {SUBMIT_SETTINGS_FORM}
          </button>
        </form>
      )}
    </>
  );
}
