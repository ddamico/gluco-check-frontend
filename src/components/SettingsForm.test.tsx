import React from "react";
import {
  cleanup,
  findByRole,
  render,
  act,
  prettyDOM,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { axe, toHaveNoViolations } from "jest-axe";
import { BloodGlucoseUnits } from "../lib/enums";
import SettingsForm from "./SettingsForm";
import { SUBMIT_SETTINGS_FORM } from "../lib/strings";

expect.extend(toHaveNoViolations);

afterEach(() => {
  jest.clearAllMocks();
  cleanup();
});

describe("SettingsForm component", () => {
  const mockNsUrl = "https://example.com";
  const mockNsToken = "token123";
  const mockGlucoseUnits = BloodGlucoseUnits.mgdl;
  const mockOnSubmit = jest.fn();

  it("renders the component", async () => {
    const { container, findByText, findByLabelText } = render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        onSubmit={mockOnSubmit}
      />
    );
    const submitButton = await findByText(SUBMIT_SETTINGS_FORM);
    const tokenField = await findByLabelText("Nightscout Token");
    const urlField = await findByLabelText("Nightscout URL");
    const glucoseUnitsSelect = await findByLabelText("Glucose Units");

    expect(container.firstChild).toMatchSnapshot();
    expect(submitButton).not.toBeEnabled();
    expect(tokenField).toHaveValue(mockNsToken);
    expect(urlField).toHaveValue(mockNsUrl);
    expect(glucoseUnitsSelect).toHaveValue(mockGlucoseUnits);
  });

  it("submits the form with the correct values", async () => {
    expect.assertions(1);
    mockOnSubmit.mockImplementationOnce(async (data) => {
      expect(data).toMatchInlineSnapshot(`
        Object {
          "glucoseUnit": "mg/dl",
          "nightscoutToken": "token123token",
          "nightscoutUrl": "https://example.com",
        }
      `);
    });

    const { findByText, findByLabelText } = render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        onSubmit={mockOnSubmit}
      />
    );
    const submitButton = await findByText(SUBMIT_SETTINGS_FORM);
    const tokenField = await findByLabelText("Nightscout Token");

    act(() => {
      userEvent.type(tokenField, "token");
    });
    act(() => {
      userEvent.click(submitButton);
    });
  });

  it("whatever, error", async () => {
    mockOnSubmit.mockImplementationOnce(async () => {
      throw new Error("To err is human.");
    });

    const { findByText, findByLabelText } = render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        onSubmit={mockOnSubmit}
      />
    );
    const submitButton = await findByText(SUBMIT_SETTINGS_FORM);
    const tokenField = await findByLabelText("Nightscout Token");

    act(() => {
      userEvent.type(tokenField, "token");
    });
    act(() => {
      userEvent.click(submitButton);
    });
  });

  it("has no axe violations", async () => {
    const { container } = render(
      <SettingsForm
        nightscoutToken={mockNsToken}
        nightscoutUrl={mockNsUrl}
        glucoseUnit={mockGlucoseUnits}
        onSubmit={mockOnSubmit}
      />
    );
    expect(await axe(container)).toHaveNoViolations();
  });
});
