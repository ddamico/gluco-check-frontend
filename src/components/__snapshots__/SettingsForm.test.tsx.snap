// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsForm component renders the component 1`] = `
<form>
  <label>
    Nightscout URL
    <input
      name="nightscoutUrl"
      value="https://example.com"
    />
  </label>
  <br />
  <label>
    Nightscout Token
    <input
      name="nightscoutToken"
      value="token123"
    />
  </label>
  <br />
  <label>
    Glucose Units
    <select
      name="glucoseUnit"
    >
      <option
        id="mg/dl"
        selected=""
      >
        mg/dl
      </option>
      <option
        id="mmol/L"
      >
        mmol/L
      </option>
    </select>
  </label>
  <br />
  <button
    disabled=""
    type="submit"
  >
    Save Settings
  </button>
</form>
`;

exports[`SettingsForm component submits the form with the correct values 1`] = `
<form>
  <label>
    Nightscout URL
    <input
      name="nightscoutUrl"
      value="https://example.com"
    />
  </label>
  <br />
  <label>
    Nightscout Token
    <input
      name="nightscoutToken"
      value="token123"
    />
  </label>
  <br />
  <label>
    Glucose Units
    <select
      name="glucoseUnit"
    >
      <option
        id="mg/dl"
        selected=""
      >
        mg/dl
      </option>
      <option
        id="mmol/L"
      >
        mmol/L
      </option>
    </select>
  </label>
  <br />
  <button
    disabled=""
    type="submit"
  >
    Save Settings
  </button>
</form>
`;
